docs/spec/staking/end_block.md

```diff
@@ -1,68 +1,40 @@
 # End-Block 
 
-## Unbonding Validator Queue
+## Validator Set Changes
 
-For all unbonding validators that have finished their unbonding period, this switches their validator.Status
-from sdk.Unbonding to sdk.Unbonded if they still have any delegation left.  Otherwise, it deletes it from state.
+The Tendermint validator set is be updated by state transitions that run at the
+end of every block. Operations are as following:
 
-```golang
-validatorQueue(currTime time.Time):
-    // unbonding validators are in ordered queue from oldest to newest
-    for all unbondingValidators whose CompleteTime < currTime:
-        validator = GetValidator(unbondingValidator.ValidatorAddr)
-        if validator.DelegatorShares == 0 {
-            RemoveValidator(unbondingValidator)
-        } else {
-            validator.Status = sdk.Unbonded
-            SetValidator(unbondingValidator)
-        }
-    return
-```
+ - The new validator set is taken as the top `params.MaxValidators` number of
+   validators retrieved from the ValidatorsByPower index
+ - The previous validator set is compared with the new validator set 
+   - missing validators begin unbonding
+   - new validator are instantly bonded
 
-## Validator Set Changes
+in all cases any validators leaving or entering the bonded validator set or
+changing balances and staying within the bonded validator set incure an update
+message which is passed back to Tendermint.
 
-The Tendermint validator set may be updated by state transitions that run at
-the end of every block. The Tendermint validator set may be changed by
-validators either being jailed due to inactivity/unexpected behaviour (covered
-in slashing) or changed in validator power. Determining which validator set
-changes must be made occurs during staking transactions (and slashing
-transactions) - during end-block the already accounted changes are applied and
-the changes cleared
+## Queues 
 
-```golang
-EndBlock() ValidatorSetChanges
-    vsc = GetValidTendermintUpdates()
-    ClearTendermintUpdates()
-    return vsc
-```
+### Unbonding Validators
 
-## CompleteUnbonding
+Each block the validator queue is to be checked for mature unbonding
+validators.  For all unbonding validators that have finished their unbonding
```

Comment:
Is there a full flow here that describes the validator lifecycle? Would be hugely helpful for understanding this.

Reply:
Validator lifecycle is a great idea! - I think it goes more in the explanatory documentation rather than the spec... will open an issue

Reply:
I agree, however, the intention of this PR was primarily to update existing spec sections... Although I think the validator lifecycle is a valueable documentation which should be linked to in this spec, I don't believe it should live in the _spec_ but instead in a new documentation section which is `EXPLANATION` oriented rather than `REFERENCE` oriented (aka. the primary intention of this spec) - as per https://www.divio.com/blog/documentation/.  Seeing as this new document is distinguished from the spec and that the intention of this PR is update the spec, I see it most appropriate to include this new document in a new PR - here is the reference issue:
https://github.com/cosmos/cosmos-sdk/issues/3325